### PR TITLE
[SYCL] Implement tracing for in-memory kernel and program cache

### PR DESCRIPTION
**Pull request by uditagarwal97**
I'm working on implementing cache eviction and it gets difficult to debug/write test cases without tracing. So, this PR implements tracing for in-memory kernel and program cache.

### CODE REVIEW

File: sycl/source/detail/kernel_program_cache.hpp [hunk 1]
```diff
@@ -176,6 +179,49 @@ class KernelProgramCache {
 
   void setContextPtr(const ContextPtr &AContext) { MParentContext = AContext; }
 
+  /* Sends message to std:cerr stream when SYCL_CACHE_TRACE environemnt is
+   * set.*/
```
Comment (aelovikov-intel) [inline sycl/source/detail/kernel_program_cache.hpp:183]: Use C++ (`//`) comments.

File: sycl/source/detail/kernel_program_cache.hpp [hunk 2]
```diff
@@ -176,6 +179,49 @@ class KernelProgramCache {
...
+  static inline void traceProgram(const std::string &Msg,
+                                  const ProgramCacheKeyT &CacheKey) {
+    static const bool traceEnabled =
+        SYCLConfig<SYCL_CACHE_TRACE>::isTraceInMemCache();
+    if (traceEnabled) {
```
Comment (aelovikov-intel) [inline sycl/source/detail/kernel_program_cache.hpp:188]: LLVM style guide requires us to early return here.

File: sycl/test-e2e/KernelAndProgram/trace_kernel_program_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+// There should be no tracing output when SYCL_CACHE_IN_MEM is not set
+// or SYCL_CACHE_TRACE is set to 0.
+
+// RUN: env SYCL_CACHE_IN_MEM=0 %{run} %t.out \
+// RUN: | FileCheck --allow-empty --check-prefix=CHECK-NO-TRACE %s
+// RUN: env SYCL_CACHE_TRACE=0 %{run} %t.out \
+// RUN: | FileCheck --allow-empty --check-prefix=CHECK-NO-TRACE %s
+
+// CHECK-NO-TRACE-NOT: [In-Memory Cache]{{.*}}
```
Comment (aelovikov-intel) [inline sycl/test-e2e/KernelAndProgram/trace_kernel_program_cache.cpp:13]: I think `--check-not "Cache"` would work just fine

File: sycl/test-e2e/KernelAndProgram/trace_kernel_program_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+// RUN: env SYCL_CACHE_TRACE=2 %{run} %t.out 2> %t.trace
+// RUN: FileCheck %s --input-file=%t.trace --check-prefix=CHECK-CACHE-TRACE
```
Comment (aelovikov-intel) [inline sycl/test-e2e/KernelAndProgram/trace_kernel_program_cache.cpp:16]: Do we need yet another RUN line to check some persistent caching?

Comment (uditagarwal97) [reply sycl/test-e2e/KernelAndProgram/trace_kernel_program_cache.cpp]: Nope. This test is only for testing in-memory cache tracing.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers the PR is ready to be merged.

**Pull request merged**